### PR TITLE
[lldb] Fix `FindDirectNestedType` not working with class templates

### PR DESCRIPTION
**Pull request by Endilll**
This patch attempts to fix lookup in class template specialization.

The first fixed problem is that during type lookup `DeclContextGetName` have been dropping template arguments. So when such a name was compared against a name in `DW_AT_name`, which contains template arguments, false mismatches have been occurring.

The second fixed problem is that LLDB's printing policy hasn't been matching Clang's printing policy when it comes to integral non-type template arguments. This again caused some false mismatches during type lookup, because Clang puts e.g. `3U` in debug info for class specializations, but LLDB has been expecting just `3`. This patch brings printing policy in line with what Clang does.

### CODE REVIEW

**Comment by Endilll** [comment]
This PR is a draft because new test, which is simplified `llvm::PointerIntPair` and `llvm::PointerIntPairInfo`, doesn't pass even with the proposed fix.

File: lldb/test/API/python_api/type/TestTypeList.py [hunk 1]
```diff
@@ -150,6 +150,23 @@ def test(self):
         invalid_type = task_type.FindDirectNestedType(None)
         self.assertFalse(invalid_type)
 
+        # Check that FindDirectNestedType works with types from AST
+        pointer = frame0.FindVariable("pointer")
+        pointer_type = pointer.GetType()
+        self.assertTrue(pointer_type)
+        self.DebugSBType(pointer_type)
+        pointer_info_type = pointer_type.template_args[0]
```
Comment (Michael137) [inline lldb/test/API/python_api/type/TestTypeList.py:158]:
```suggestion
        pointer_info_type = pointer_type.template_args[1]
```

then the tests works :)

Comment (Endilll) [reply lldb/test/API/python_api/type/TestTypeList.py]: You're absolutely right. Thank you!

**Review by Michael137 — APPROVED** [review]
LGTM

Can you add a bit more detail into the commit message and PR description about what the underlying problem was? I.e., `DeclContextGetName` used to return type names without template parameters which `FindTypes` would then compare against the `DW_AT_name`s from DWARF. But the names in DWARF *do* contain template parameters, causing the lookups to fail.

**Comment by JDevlieghere** [comment]
This breaks `TestDataFormatterLibcxxChrono.py` on macOS: 

https://green.lab.llvm.org/green/view/LLDB/job/lldb-cmake/66400/
https://green.lab.llvm.org/green/view/LLDB/job/as-lldb-cmake/16178/

CC @rastogishubham

**Comment by Endilll** [comment]
Trying to take a look, but https://green.lab.llvm.org is extraordinary slow for me.

**Comment by JDevlieghere** [comment]
This is what the failure looks like:

```
======================================================================
FAIL: test_with_run_command_dsym (TestDataFormatterLibcxxChrono.LibcxxChronoDataFormatterTestCase)
   Test that that file and class static variables display correctly.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jonas/llvm/llvm-project/lldb/packages/Python/lldbsuite/test/lldbtest.py", line 1676, in test_method
    return attrvalue(self)
  File "/Users/jonas/llvm/llvm-project/lldb/packages/Python/lldbsuite/test/decorators.py", line 150, in wrapper
    return func(*args, **kwargs)
  File "/Users/jonas/llvm/llvm-project/lldb/test/API/functionalities/data-formatter/data-formatter-stl/libcxx/chrono/TestDataFormatterLibcxxChrono.py", line 35, in test_with_run_command
    self.expect(
  File "/Users/jonas/llvm/llvm-project/lldb/packages/Python/lldbsuite/test/lldbtest.py", line 2447, in expect
    self.fail(log_msg)
AssertionError: Ran command:
"frame variable ss_tp"

Got output:
(std::chrono::time_point<std::chrono::system_clock, std::chrono::duration<long long> >) ss_tp = {
  __d_ = (__rep_ = 0)
}

Expecting sub string: "ss_tp = date/time=1970-01-01T00:00:00Z timestamp=0 s" (was not found)
```

**Comment by Endilll** [comment]
@JDevlieghere I have an idea what have caused this failure, but I guess I have to enable libcxx to replicate it locally.

**Comment by JDevlieghere** [comment]
Great. Let me know if you need help reproducing stuff on macOS.

**Comment by rastogishubham** [comment]
I am going to revert this patch for now, to make sure that green dragon is up and running again

**Comment by rastogishubham** [comment]
Hi @Endilll this patch has been reverted with 831ba9540089350b740c5db61159fe23ab6872d3 to make sure the lldb bots are green.

**Pull request merged**